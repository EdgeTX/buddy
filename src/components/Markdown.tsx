import React from "react";
import ReactMarkdown from "react-markdown";
import Typography from "@mui/material/Typography";
import Link from "@mui/material/Link";
import Table from "@mui/material/Table";
import TableContainer from "@mui/material/TableContainer";
import Paper from "@mui/material/Paper";
import { TableHead, TableRow, TableCell, TableBody } from "@mui/material";
import { Variant } from "@mui/material/styles/createTypography";

const MarkdownParagraph: React.FC = (props) => {
  return <Typography>{props.children}</Typography>;
};

const MarkdownHeading: React.FC<{ level: number }> = (props) => {
  let variant: Variant;
  switch (props.level) {
    case 1:
      variant = "h5";
      break;
    case 2:
      variant = "h6";
      break;
    case 3:
      variant = "subtitle1";
      break;
    case 4:
      variant = "subtitle2";
      break;
    default:
      variant = "h6";
      break;
  }
  return (
    <Typography sx={{ marginTop: 2 }} gutterBottom variant={variant}>
      {props.children}
    </Typography>
  );
};

const MarkdownListItem: React.FC = (props) => {
  return (
    <li>
      <Typography component="span">{props.children}</Typography>
    </li>
  );
};

const MarkdownTable: React.FC = (props) => {
  return (
    <TableContainer component={Paper}>
      <Table size="small" aria-label="a dense table">
        {props.children}
      </Table>
    </TableContainer>
  );
};

const MarkdownTableCell: React.FC = (props) => {
  return (
    <TableCell>
      <Typography>{props.children}</Typography>
    </TableCell>
  );
};

const MarkdownTableRow: React.FC = (props) => {
  return <TableRow>{props.children}</TableRow>;
};

const MarkdownTableBody: React.FC = (props) => {
  return <TableBody>{props.children}</TableBody>;
};

const MarkdownTableHead: React.FC = (props) => {
  return <TableHead>{props.children}</TableHead>;
};

const Markdown: typeof ReactMarkdown = ((props) => {
  return (
    <ReactMarkdown
      components={
        {
          heading: MarkdownHeading,
          paragraph: MarkdownParagraph,
          link: Link,
          listItem: MarkdownListItem,
          table: MarkdownTable,
          tableHead: MarkdownTableHead,
          tableBody: MarkdownTableBody,
          tableRow: MarkdownTableRow,
          tableCell: MarkdownTableCell,
        } as any
      }
      {...props}
    />
  );
}) as typeof ReactMarkdown;

export default Markdown;
