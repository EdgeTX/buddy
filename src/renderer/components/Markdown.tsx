import React from "react";
import ReactMarkdown from "markdown-to-jsx";
import Typography from "@mui/material/Typography";
import Link from "@mui/material/Link";
import Table from "@mui/material/Table";
import TableContainer from "@mui/material/TableContainer";
import Paper from "@mui/material/Paper";
import { TableHead, TableRow, TableCell, TableBody } from "@mui/material";
import { Variant } from "@mui/material/styles/createTypography";

const MarkdownParagraph: React.FC = ({ children }) => (
  <Typography>{children}</Typography>
);

const MarkdownHeading: React.FC<{ level: number }> = ({ children, level }) => {
  let variant: Variant;
  switch (level) {
    case 1:
      variant = "h5";
      break;
    case 2:
      variant = "h6";
      break;
    case 3:
      variant = "subtitle1";
      break;
    case 4:
      variant = "subtitle2";
      break;
    default:
      variant = "h6";
      break;
  }
  return (
    <Typography sx={{ marginTop: 2 }} gutterBottom variant={variant}>
      {children}
    </Typography>
  );
};

const MarkdownListItem: React.FC = ({ children }) => (
  <li>
    <Typography component="span">{children}</Typography>
  </li>
);

const MarkdownTable: React.FC = ({ children }) => (
  <TableContainer component={Paper}>
    <Table size="small" aria-label="a dense table">
      {children}
    </Table>
  </TableContainer>
);

const MarkdownTableCell: React.FC = ({ children }) => (
  <TableCell>
    <Typography>{children}</Typography>
  </TableCell>
);

const MarkdownTableRow: React.FC = ({ children }) => (
  <TableRow>{children}</TableRow>
);

const MarkdownTableBody: React.FC = ({ children }) => (
  <TableBody>{children}</TableBody>
);

const MarkdownTableHead: React.FC = ({ children }) => (
  <TableHead>{children}</TableHead>
);

const Markdown: typeof ReactMarkdown = ((props) => (
  <ReactMarkdown
    options={{
      overrides: {
        h1: {
          component: MarkdownHeading,
          props: {
            level: 1,
          },
        },
        h2: {
          component: MarkdownHeading,
          props: {
            level: 2,
          },
        },
        h3: {
          component: MarkdownHeading,
          props: {
            level: 3,
          },
        },
        h4: {
          component: MarkdownHeading,
          props: {
            level: 4,
          },
        },
        h5: {
          component: MarkdownHeading,
          props: {
            level: 5,
          },
        },
        h6: {
          component: MarkdownHeading,
          props: {
            level: 6,
          },
        },
        p: MarkdownParagraph,
        a: Link,
        li: MarkdownListItem,
        table: MarkdownTable,
        th: MarkdownTableHead,
        tbody: MarkdownTableBody,
        tr: MarkdownTableRow,
        td: MarkdownTableCell,
      },
    }}
    // eslint-disable-next-line react/jsx-props-no-spreading
    {...props}
  />
)) as typeof ReactMarkdown;

export default Markdown;
