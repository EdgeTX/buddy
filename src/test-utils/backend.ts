/* eslint-disable import/prefer-default-export */
import {
  DocumentNode,
  execute,
  ExecutionArgs,
  ExecutionResult,
  OperationDefinitionNode,
} from "graphql";
import { createContext, schema, UsbApi, FileSystemApi } from "shared/backend";
import { Context } from "shared/backend/context";
import * as dfuApi from "shared/backend/services/dfu";

export type PromiseOrValue<T> = Promise<T> | T;

type QueryArgs = { query: DocumentNode; variables?: Record<string, unknown> };
type MutationArgs = {
  mutation: DocumentNode;
  variables?: Record<string, unknown>;
};
type QueryFunction = (args: QueryArgs) => PromiseOrValue<ExecutionResult>;
type MutationFunction = (args: MutationArgs) => ReturnType<QueryFunction>;

type Executor = {
  query: QueryFunction;
  mutate: MutationFunction;
  context: Context;
};

const getOperationName = (document: DocumentNode): string | undefined =>
  document.definitions.find(
    (def): def is OperationDefinitionNode => def.kind === "OperationDefinition"
  )?.name?.value;

export const createExecutor = ({
  usb = {} as UsbApi,
  fileSystem = {} as FileSystemApi,
  dfu = {} as typeof dfuApi,
} = {}): Executor => {
  const contextGenerator = createContext({
    usb,
    fileSystem,
    dfu,
  });

  const executeQuery: QueryFunction = ({ query, variables }) => {
    const args: ExecutionArgs = {
      schema,
      contextValue: contextGenerator(),
      variableValues: variables,
      operationName: getOperationName(query),
      document: query,
    };

    return execute(args);
  };

  return {
    query: executeQuery,
    mutate: ({ mutation, variables }) =>
      executeQuery({ query: mutation, variables }),
    context: contextGenerator(),
  };
};
