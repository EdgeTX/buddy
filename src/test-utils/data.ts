export const exampleReleasesList = [
  {
    id: "nightly",
    name: "nightly",
    isPrerelease: true,
  },
  {
    id: "v2.5.0",
    name: 'EdgeTX "Dauntless" 2.5.0',
    isPrerelease: false,
  },
  {
    id: "v2.5.0-rc3",
    name: 'EdgeTX "Dauntless" 2.5.0-RC3',
    isPrerelease: true,
  },
  {
    id: "v2.5.0-rc2",
    name: 'EdgeTX "Dauntless" 2.5.0-RC2',
    isPrerelease: true,
  },
  {
    id: "v2.5.0-rc1",
    name: 'EdgeTX "Dauntless" 2.5.0-RC1',
    isPrerelease: true,
  },
  {
    id: "v2.4.0",
    name: 'EdgeTX "Endeavour" 2.4.0',
    isPrerelease: false,
  },
  {
    id: "v2.4.0-rc4",
    name: "EdgeTX Endeavour (2.4.0) RC4",
    isPrerelease: true,
  },
  {
    id: "v2.4.0-rc3",
    name: "EdgeTX Endeavour (2.4.0) RC3",
    isPrerelease: true,
  },
  {
    id: "v2.4.0-rc2",
    name: "EdgeTX Endeavour (2.4.0) RC2",
    isPrerelease: true,
  },
  {
    id: "v2.4.0-rc1",
    name: "2.4.0-rc1",
    isPrerelease: true,
  },
];

export const exampleTargetsList = [
  {
    code: "nv14",
    id: "nv14-50406116",
    name: "Flysky NV14",
  },
  {
    code: "x10",
    id: "x10-50406116",
    name: "Frsky Horus X10",
  },
  {
    code: "x10-access",
    id: "x10-access-50406116",
    name: "Frsky Horus X10 Access",
  },
  {
    code: "x12s",
    id: "x12s-50406116",
    name: "Frsky Horus X12s",
  },
  {
    code: "x7",
    id: "x7-50406116",
    name: "Frsky QX7",
  },
  {
    code: "x7-access",
    id: "x7-access-50406116",
    name: "Frsky QX7 Access",
  },
  {
    code: "x9d",
    id: "x9d-50406116",
    name: "Frsky X9D",
  },
  {
    code: "x9dp",
    id: "x9dp-50406116",
    name: "Frsky X9D Plus",
  },
  {
    code: "x9dp2019",
    id: "x9dp2019-50406116",
    name: "Frsky X9D Plus 2019",
  },
  {
    code: "x9lite",
    id: "x9lite-50406116",
    name: "Frsky X9 Lite",
  },
  {
    code: "x9lites",
    id: "x9lites-50406116",
    name: "Frsky X9 Lite S",
  },
  {
    code: "xlite",
    id: "xlite-50406116",
    name: "Frsky X-Lite",
  },
  {
    code: "xlites",
    id: "xlites-50406116",
    name: "Frsky X-Lite S",
  },
  {
    code: "t8",
    id: "t8-50406116",
    name: "Jumper T8",
  },
  {
    code: "t12",
    id: "t12-50406116",
    name: "Jumper T12",
  },
  {
    code: "t16",
    id: "t16-50406116",
    name: "Jumper T16",
  },
  {
    code: "t18",
    id: "t18-50406116",
    name: "Jumper T18",
  },
  {
    code: "tlite",
    id: "tlite-50406116",
    name: "Jumper T-Lite",
  },
  {
    code: "tx12",
    id: "tx12-50406116",
    name: "Radiomaster TX12",
  },
  {
    code: "tx16s",
    id: "tx16s-50406116",
    name: "Radiomaster TX16s",
  },
];

export const exampleDevices = [
  {
    id: "1",
    serialNumber: "375897983475",
    productName: "STEM32 Bootloader",
    vendorId: "0x3453",
    productId: "0x4358",
  },
  {
    id: "2",
    serialNumber: "2435883",
    productName: "Radio 2",
    vendorId: "0x5467",
    productId: "0x4668",
  },
  {
    id: "3",
    serialNumber: "2435883",
    productName: "Radio 3",
    vendorId: "0x5467",
    productId: "0x4668",
  },
  {
    id: "4",
    serialNumber: "2435883",
    productName: "Radio 4",
    vendorId: "0x5467",
    productId: "0x4668",
  },
  {
    id: "5",
    serialNumber: "375897983475",
    productName: "STEM32 Bootloader 2",
    vendorId: "0x3453",
    productId: "0x4358",
  },
  {
    id: "6",
    serialNumber: "375897983475",
    productName: "STEM32 Bootloader 3",
    vendorId: "0x3453",
    productId: "0x4358",
  },
  {
    id: "6",
    serialNumber: "375897983475",
    productName: "STEM32 Bootloader 4",
    vendorId: "0x3453",
    productId: "0x4358",
  },
];

export const examplePrs = [
  {
    headCommitId: "217c02e6e06b4500edbb0eca99b5d1d077111aab",
    id: "1333",
    name: "wimalopaan:wm1333",
  },
  {
    headCommitId: "2fceb40b70432378d9e0693e0d3939221799ba86",
    id: "984",
    name: "EdgeTX:crsf-support-lua-command-id",
  },
  {
    headCommitId: "621a6833106e744b7e4b616ad6218af7c7d47e19",
    id: "1009",
    name: "jfrickmann:LCD",
  },
  {
    headCommitId: "ceafdea344cd9bc021dd083fe4a487393edb1458",
    id: "980",
    name: "EdgeTX:nv14-bigger-controls",
  },
  {
    headCommitId: "c05c7066af63dfd30df28fbf1355ef19eca0f574",
    id: "1115",
    name: "stronnag:add_MP_header_to_Telem_Mirror",
  },
];

export const examplePrCommits = [
  {
    id: "217c02e6e06b4500edbb0eca99b5d1d077111aab",
  },
  {
    id: "b9319c5428aab118b9473e98fcde996355333c8c",
  },
];

export const exampleCloudbuildTargets = {
  releases: {
    nightly: {
      sha: "06712d9b977ee19170a3abc3c183b2f17bd8fa35",
      remote: {
        url: "https://api.github.com/repos/EdgeTX/edgetx/git/ref/tags/nightly",
        json_path: "",
      },
    },
    "v2.8.4": {
      sha: "da28e356449e54c57f0e5e356bd5ec5709128ff7",
      remote: null,
      exclude_targets: ["tlitef4"],
    },
  },
  flags: {
    language: {
      build_flag: "TRANSLATIONS",
      values: [
        "EN",
        "CZ",
        "DA",
        "DE",
        "EN",
        "ES",
        "FI",
        "FR",
        "IT",
        "PT",
        "SK",
        "SE",
        "PL",
        "HU",
        "NL",
      ],
    },
  },
  tags: { colorlcd: { flags: { language: { values: ["CN", "JP", "TW"] } } } },
  targets: {
    boxer: {
      description: "RadioMaster Boxer",
      build_flags: { PCB: "X7", PCBREV: "BOXER" },
    },
    commando8: {
      description: "iFlight Commando 8",
      build_flags: { PCB: "X7", PCBREV: "COMMANDO8" },
    },
    el18: {
      description: "Flysky EL18",
      tags: ["colorlcd"],
      build_flags: { PCB: "NV14" },
    },
    lr3pro: {
      description: "BETAFPV LiteRadio 3 Pro",
      build_flags: { PCB: "X7", PCBREV: "LR3PRO" },
    },
    nv14: {
      description: "Flysky NV14",
      tags: ["colorlcd"],
      build_flags: { PCB: "NV14" },
    },
    t12: {
      description: "Jumper T12",
      build_flags: { PCB: "X7", PCBREV: "T12" },
    },
    t16: {
      description: "Jumper T16",
      tags: ["colorlcd"],
      build_flags: { PCB: "X10", PCBREV: "T16" },
    },
    t18: {
      description: "Jumper T18",
      tags: ["colorlcd"],
      build_flags: { PCB: "X10", PCBREV: "T18" },
    },
    t8: {
      description: "Radiomaster T8",
      build_flags: { PCB: "X7", PCBREV: "T8" },
    },
    tlite: {
      description: "Jumper T-Lite",
      build_flags: { PCB: "X7", PCBREV: "TLITE" },
    },
    tlitef4: {
      description: "Jumper T-Lite (F4 MCU)",
      build_flags: { PCB: "X7", PCBREV: "TLITEF4" },
    },
    tpro: {
      description: "Jumper T-Pro",
      build_flags: { PCB: "X7", PCBREV: "TPRO" },
    },
    tx12: {
      description: "Radiomaster TX12MK2",
      build_flags: { PCB: "X7", PCBREV: "TX12MK2" },
    },
    tx16s: {
      description: "RadioMaster TX16S",
      tags: ["colorlcd"],
      build_flags: { PCB: "X10", PCBREV: "TX16S" },
    },
    x10: {
      description: "FrSky Horus X10",
      tags: ["colorlcd"],
      build_flags: { PCB: "X10" },
    },
    "x10-access": {
      description: "FrSky Horus X10 Access",
      tags: ["colorlcd"],
      build_flags: { PCB: "X10", PCBREV: "EXPRESS", PXX1: "YES" },
    },
    x12s: {
      description: "FrSky Horus X12S",
      tags: ["colorlcd"],
      build_flags: { PCB: "X12S" },
    },
    x7: { description: "FrSky QX7", build_flags: { PCB: "X7" } },
    "x7-access": {
      description: "FrSky QX7 Access",
      build_flags: { PCB: "X7", PCBREV: "ACCESS" },
    },
    x9d: { description: "FrSky X9D", build_flags: { PCB: "X9D" } },
    x9dp: {
      description: "FrSky X9D Plus",
      build_flags: { PCB: "X9D+", PCBREV: "ACCESS" },
    },
    x9dp2019: {
      description: "FrSky X9D Plus 2019",
      build_flags: { PCB: "X9D+", PCBREV: "2019" },
    },
    x9e: { description: "FrSky X9E", build_flags: { PCB: "X9E" } },
    x9lite: { description: "FrSky X9 Lite", build_flags: { PCB: "X9LITE" } },
    x9lites: {
      description: "FrSky X9 Lite S",
      build_flags: { PCB: "X9LITES" },
    },
    xlite: { description: "FrSky X-Lite", build_flags: { PCB: "XLITE" } },
    xlites: { description: "FrSky X-Lite S", build_flags: { PCB: "XLITES" } },
    zorro: {
      description: "RadioMaster Zorro",
      build_flags: { PCB: "X7", PCBREV: "ZORRO" },
    },
  },
};
