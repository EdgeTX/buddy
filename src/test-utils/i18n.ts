import i18n from "i18next";

void i18n.init({
  lng: "en",
  fallbackLng: "en",

  // have a common namespace used around the full app
  ns: ["common"],
  defaultNS: "common",

  interpolation: {
    escapeValue: false, // not needed for react!!
  },

  resources: { en: { common: {} } },
});

export default i18n;
