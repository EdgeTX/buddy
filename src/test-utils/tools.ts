import dirTree from "directory-tree";

export const directorySnapshot = (path: string): string => {
  const tree = dirTree(path, { normalizePath: true, attributes: ["size"] });
  tree.name = ".";

  const treeString = JSON.stringify(tree, undefined, 2);
  return treeString.replaceAll(path, ".");
};

export const waitForStageCompleted = async <
  T extends {
    stages: { [K in S]?: { completed: boolean; error?: string | null } | null };
  },
  S extends string
>(
  queue: AsyncIterator<T, any, undefined>,
  stage: S
): Promise<void> => {
  while (true) {
    const update = await queue.next();
    if (!update.done) {
      const job = update.value;
      if (job.stages[stage]?.completed) {
        return;
      }

      if (job.stages[stage]?.error) {
        throw new Error(job.stages[stage]?.error ?? "");
      }
    }
  }
};
