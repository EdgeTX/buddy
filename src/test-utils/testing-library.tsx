import React from "react";
import { fireEvent, render as origRender } from "@testing-library/react";
import AntProvider from "renderer/AntProvider";
import { I18nextProvider } from "react-i18next";
import i18n from "./i18n";

export const render = (children: React.ReactNode) => {
  const result = origRender(
    <I18nextProvider i18n={i18n}>
      <AntProvider>{children}</AntProvider>)
    </I18nextProvider>
  );
  return {
    ...result,
    rerender: (c: React.ReactNode) =>
      result.rerender(
        <I18nextProvider i18n={i18n}>
          <AntProvider>{c}</AntProvider>
        </I18nextProvider>
      ),
  };
};

export const openAntDropdown = (element: HTMLElement) => {
  fireEvent.keyDown(element, {
    key: "Enter",
    code: 13,
    charCode: 13,
  });
};
