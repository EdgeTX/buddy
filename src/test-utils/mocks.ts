import { MockedResponse } from "@apollo/client/testing";
import gql from "graphql-tag";
import { times } from "shared/tools";
import {
  exampleCloudbuildTargets,
  exampleDevices,
  examplePrCommits,
  examplePrs,
  exampleReleasesList,
  exampleTargetsList,
} from "test-utils/data";

export const firmwaresQuery = (delay = 100): MockedResponse => ({
  request: {
    query: gql`
      query Releases {
        edgeTxReleases {
          id
          name
          isPrerelease
        }
      }
    `,
  },
  result: {
    data: {
      edgeTxReleases: exampleReleasesList,
    },
  },
  delay,
});

export const targetsQuery = (delay = 1000): MockedResponse => ({
  request: {
    query: gql`
      query ReleaseTargets($releaseId: ID!) {
        edgeTxRelease(id: $releaseId) {
          id
          firmwareBundle {
            id
            targets {
              id
              code
              name
            }
          }
        }
      }
    `,
    variables: {
      releaseId: "v2.5.0",
    },
  },
  result: {
    data: {
      edgeTxRelease: {
        __typename: "EdgeTxRelease",
        id: "v2.5.0",
        firmwareBundle: {
          id: "",
          targets: exampleTargetsList,
        },
      },
    },
  },
  delay,
});

const releaseDescription = times(10)
  .map(
    () => `
My release, it is great, it has lots of things.

- This thing
- and this
- and this.

[View it here](https://google.com)
`
  )
  .join("");

export const firmwareReleaseDescriptionQuery = (
  delay = 2000
): MockedResponse => ({
  request: {
    query: gql`
      query FirmwareReleaseDescription($releaseId: ID!) {
        edgeTxRelease(id: $releaseId) {
          id
          description
        }
      }
    `,
    variables: {
      releaseId: "v2.5.0",
    },
  },
  delay,
  result: {
    data: {
      edgeTxRelease: {
        __typename: "EdgeTxRelease",
        id: "v2.5.0",
        description: releaseDescription,
      },
    },
  },
});

export const prsQuery: MockedResponse = {
  request: {
    query: gql`
      query EdgeTxPrs {
        edgeTxPrs {
          id
          name
          headCommitId
        }
      }
    `,
  },
  result: {
    data: {
      edgeTxPrs: examplePrs,
    },
  },
};

export const prCommitsQuery: MockedResponse = {
  request: {
    query: gql`
      query EdgeTxPrCommits($prId: ID!) {
        edgeTxPr(id: $prId) {
          id
          commits {
            id
          }
        }
      }
    `,
    variables: {
      prId: examplePrs[0]?.id,
    },
  },
  result: {
    data: {
      edgeTxPr: {
        __typename: "EdgeTxPr",
        id: examplePrs[0]?.id,
        commits: examplePrCommits,
      },
    },
  },
};

export const prCommitBuildQuery: MockedResponse = {
  request: {
    query: gql`
      query EdgeTxPrCommitBuild($prId: ID!, $commitId: ID!) {
        edgeTxPr(id: $prId) {
          id
          commit(id: $commitId) {
            id
            firmwareBundle {
              id
              targets {
                id
                code
                name
              }
            }
          }
        }
      }
    `,
    variables: {
      prId: examplePrs[0]?.id,
      commitId: examplePrs[0]?.headCommitId,
    },
  },
  result: {
    data: {
      edgeTxPr: {
        __typename: "EdgeTxPr",
        id: examplePrs[0]?.id,
        commit: {
          firmwareBundle: {
            id: "135237194",
            targets: exampleTargetsList,
          },
          id: examplePrs[0]?.headCommitId,
        },
      },
    },
  },
};

export const prCommitBuildNotAvailableQuery: MockedResponse = {
  request: {
    query: gql`
      query EdgeTxPrCommitBuild($prId: ID!, $commitId: ID!) {
        edgeTxPr(id: $prId) {
          id
          commit(id: $commitId) {
            id
            firmwareBundle {
              id
              targets {
                id
                code
                name
              }
            }
          }
        }
      }
    `,
    variables: {
      prId: examplePrs[0]?.id,
      commitId: examplePrCommits[1]?.id,
    },
  },
  result: {
    data: {
      edgeTxPr: {
        __typename: "EdgeTxPr",
        id: examplePrs[0]?.id,
        commit: {
          firmwareBundle: null,
          id: examplePrCommits[1]?.id,
        },
      },
    },
  },
};

export const prDescriptionQuery: MockedResponse = {
  request: {
    query: gql`
      query PrDescription($prId: ID!) {
        edgeTxPr(id: $prId) {
          id
          title
          description
        }
      }
    `,
    variables: {
      prId: examplePrs[0]?.id,
    },
  },
  result: {
    data: {
      edgeTxPr: {
        __typename: "EdgeTxPr",
        id: examplePrs[0]?.id,
        title: "Some pr title",
        description: "- My Pr description",
      },
    },
  },
};

export const devicesQuery = (
  delay = 2000,
  devices = exampleDevices
): MockedResponse => ({
  request: {
    query: gql`
      query Devices {
        flashableDevices {
          id
          productName
          serialNumber
          vendorId
          productId
        }
      }
    `,
  },
  delay,
  result: {
    data: {
      flashableDevices: devices,
    },
  },
});

export const deviceQuery = (delay = 2000): MockedResponse => ({
  request: {
    query: gql`
      query DeviceInfo($deviceId: ID!) {
        flashableDevice(id: $deviceId) {
          id
          productName
          serialNumber
          vendorId
          productId
        }
      }
    `,
    variables: {
      deviceId: "1",
    },
  },
  delay,
  result: {
    data: {
      flashableDevice: exampleDevices[4],
    },
  },
});

export const firmwareReleaseInfoQuery = (delay = 1000): MockedResponse => ({
  request: {
    query: gql`
      query ReleaseInfo($version: ID!, $target: ID!) {
        edgeTxRelease(id: $version) {
          id
          name
          firmwareBundle {
            id
            target(code: $target) {
              id
              name
            }
          }
        }
      }
    `,
    variables: {
      version: "v2.5.0",
      target: "nv-14",
    },
  },
  result: {
    data: {
      edgeTxRelease: {
        ...exampleReleasesList[1],
        firmwareBundle: {
          id: "",
          target: exampleTargetsList[3],
        },
      },
    },
  },
  delay,
});

export const localFirmwareInfoQuery = (delay = 1000): MockedResponse => ({
  request: {
    query: gql`
      query LocalFirmwareInfo($fileId: ID!) {
        localFirmware(byId: $fileId) {
          id
          name
        }
      }
    `,
    variables: {
      fileId: "file-id-abcd",
    },
  },
  result: {
    data: {
      localFirmware: {
        id: "",
        name: "xlite-28cdb40.bin",
      },
    },
  },
  delay,
});

export const firmwarePrBuildInfoQuery = (delay = 1000): MockedResponse => ({
  request: {
    query: gql`
      query PrFirmwareInfo($prId: ID!, $commitId: ID!, $target: ID!) {
        edgeTxPr(id: $prId) {
          id
          name
          commit(id: $commitId) {
            id
            firmwareBundle {
              id
              target(code: $target) {
                id
                name
              }
            }
          }
        }
      }
    `,
    variables: {
      prId: examplePrs[0]?.id,
      commitId: examplePrs[0]?.headCommitId,
      target: "nv-14",
    },
  },
  result: {
    data: {
      edgeTxPr: {
        id: "",
        name: examplePrs[0]?.name,
        commit: {
          id: examplePrs[0]?.headCommitId,
          firmwareBundle: {
            id: "",
            target: exampleTargetsList[0],
          },
        },
      },
    },
  },
  delay,
});

export const prBuildFirmwareDataQuery: MockedResponse = {
  request: {
    query: gql`
      query PrBuildFirmwareData($prId: ID!, $commitId: ID!, $target: ID!) {
        edgeTxPr(id: $prId) {
          id
          commit(id: $commitId) {
            id
            firmwareBundle {
              id
              target(code: $target) {
                id
                base64Data
              }
            }
          }
        }
      }
    `,
    variables: {
      prId: examplePrs[0]?.id,
      commitId: examplePrs[0]?.headCommitId,
      target: "nv14",
    },
  },
  result: {
    data: {
      edgeTxPr: {
        id: "32345345",
        commit: {
          id: examplePrs[0]?.headCommitId,
          firmwareBundle: {
            id: "349345",
            target: {
              id: "nv14",
              code: "nv14",
              base64Data: Buffer.from("some-data").toString("base64"),
            },
          },
        },
      },
    },
  },
};

export const flashJobQuery = (
  jobId: string,
  error?: boolean,
  completed?: boolean
): MockedResponse => ({
  request: {
    query: gql`
      query FlashJobStatus($jobId: ID!) {
        flashJobStatus(jobId: $jobId) {
          id
          cancelled
          meta {
            firmware {
              target
              version
            }
            deviceId
          }
          stages {
            connect {
              ...FlashJobStageData
            }
            build {
              ...FlashJobStageData
            }
            download {
              ...FlashJobStageData
            }
            erase {
              ...FlashJobStageData
            }
            flash {
              ...FlashJobStageData
            }
          }
        }
      }

      fragment FlashJobStageData on FlashStage {
        started
        completed
        progress
        error
      }
    `,
    variables: {
      jobId,
    },
  },
  result: {
    data: {
      flashJobStatus: {
        id: jobId,
        cancelled: false,
        meta: {
          firmware: {
            version: "v2.5.0",
            target: "nv-14",
          },
          deviceId: "some-device-id",
        },
        stages: {
          connect: {
            started: true,
            progress: 0,
            completed: true,
          },
          download: {
            started: true,
            progress: 100,
            completed: true,
          },
          erase: {
            started: true,
            progress: 70.2,
            error: error ? "Some error" : null,
            completed: completed ?? false,
          },
          flash: {
            started: completed ?? false,
            progress: 0,
            completed: completed ?? false,
          },
        },
      },
    },
  },
});

export const cloudbuildTargets = (delay = 500): MockedResponse => ({
  request: {
    query: gql`
      query CloudTargets {
        cloudTargets {
          releases {
            id
            name
            isPrerelease
            timestamp
            excludeTargets
          }
          targets {
            id
            name
            tags
          }
          flags {
            id
            values
          }
          tags {
            id
            tagFlags {
              id
              values
            }
          }
        }
      }
    `,
  },
  result: {
    data: {
      cloudTargets: exampleCloudbuildTargets,
    },
  },
  delay,
});

const cloudbuildJobStatus = (
  release: string,
  target: string,
  flags: { name: string; value: string }[],
  build_finished?: boolean,
  error?: boolean,
  delay = 200
): MockedResponse => ({
  request: {
    query: gql`
      query CloudFirmwareStatus($params: CloudFirmwareParams!) {
        cloudFirmwareStatus(params: $params) {
          status
          downloadUrl
        }
      }
    `,
    variables: {
      params: {
        release,
        target,
        flags,
      },
    },
  },
  result: {
    data: {
      cloudFirmwareStatus: error
        ? { error: "build not found" }
        : {
            status: build_finished ? "BUILD_SUCCESS" : "BUILD_IN_PROGRESS",
            build_attempts: 1,
            artifacts: [
              {
                id: "4842244f-a1e4-4f43-93c8-8265e5951cfc",
                slug: "firmware",
                download_url: build_finished
                  ? "https://test-cloudbuild.edgetx.org/da28e356449e54c57f0e5e356bd5ec5709128ff7-fe4a260cd3251164f544654df3504a9c5d7f1e0b0d8a565941415ed4e9b8e042.bin"
                  : undefined,
                size: 492052,
                created_at: "2023-05-21T11:18:00.885181Z",
                updated_at: "2023-05-21T11:18:00.885181Z",
              },
            ],
            build_started_at: "2023-05-21T11:16:58.15048Z",
            build_ended_at: "2023-05-21T11:18:00.88349Z",
            created_at: "2023-05-21T11:16:57.877346Z",
            updated_at: "2023-05-21T11:18:00.884116Z",
          },
    },
  },
  delay,
});
