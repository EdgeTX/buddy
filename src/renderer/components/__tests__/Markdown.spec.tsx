import { describe, it, expect } from "@jest/globals";
import React from "react";
import { render } from "test-utils/testing-library";
import Markdown from "renderer/components/Markdown";

describe("<Markdown />", () => {
  it("should return the given markdown string as html", () => {
    const { asFragment } = render(
      <Markdown>
        {
          "# This is my markdown\n\n- A bullet point\n\nA paragraphy\n\n[A link](https://google.com)"
        }
      </Markdown>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
