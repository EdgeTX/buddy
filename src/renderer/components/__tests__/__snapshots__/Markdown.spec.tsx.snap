// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Markdown /> > should return the given markdown string as html 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="ant-typography"
    >
      This is my markdown
    </h1>
    <ul>
      <li>
        <article
          class="ant-typography"
        >
          A bullet point
        </article>
      </li>
    </ul>
    <p>
      A paragraphy
    </p>
    <p>
      <a
        class="ant-typography"
        href="https://google.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        A link
      </a>
    </p>
  </div>
  )
</DocumentFragment>
`;
